### ISSUE TITLE
Request for `peek` function in SCollection

### ISSUE THREAD

**Issue by sisidra** [spotify/scio#2708]
It is very common to add a counter for counting elements in SCollection, or doing validation on all SCollection elements, or similar use cases where we want to apply function on all elemets in SCollection and return the input.
Currently such use cases are server with `map`, which is slightly verbose:
```scala
scollection
  .map(element => {
    inputs.inc()
    element
  })
```
Because of that it is common to see helpers in all data repositories (or similar):
```scala
object ScioHelpers {
  implicit class RichSCollection[T: ClassTag: Coder](self: SCollection[T]) {
    def accumulateCount(counter: Counter): SCollection[T] =
      self.map { x =>
        counters.inc();
        x
      }
  }
}
```

It would be nice to add `peek` function. Name from [Java Stream API](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer-).
Scala naming would be `tap` based on [v.2.13 API](https://www.scala-lang.org/files/archive/api/2.13.x/scala/util/ChainingOps.html#tap[U](f:A=%3EU):A), and then `tap` becomes overloaded with Scio `Tap` concept... :(

**Comment by regadas** [comment]
it should be ok to have it as a `tap` actually. This way you can `tap` `IO`'s in `ScioContext`  and you can `tap` elements in `SCollection`

**Comment by nevillelyh** [comment]
`peek()` sounds to me more like the next item in a sequence, my 2c :)

**Comment by jto** [comment]
fixed by #2713
